pkg/openstack/routes.go

```diff
@@ -87,8 +124,77 @@ func (r *Routes) ListRoutes(ctx context.Context, clusterName string) ([]*cloudpr
 	}
 
 	routes := make([]*cloudprovider.Route, 0, len(router.Routes))
+
+	// detect router's private network ID for further VM ports filtering
+	r.networkIDs, err = getRouterNetworkIDs(r.network, r.os.routeOpts.RouterID)
+	if err != nil {
+		return nil, err
+	}
+
+	var sgRuleForPodCidrFound, sgRuleForNodeAddressFound, nodePortBindSecurityGroup, allowAddressPairFound bool
+out:
 	for _, item := range router.Routes {
 		nodeName, foundNode := getNodeNameByAddr(item.NextHop, nodes)
+		if foundNode {
```

Comment:
this looks like a complex code with lots of nesting. can you split it up to several functions?

Reply:
Of course